examples/tree-view/src/Node/treeSlice.js

```diff
@@ -0,0 +1,67 @@
+import { createSlice } from "@reduxjs/toolkit";
+import generateTree from "../generateTree";
+
+const { tree, idCount } = generateTree();
+
+let nextId = idCount;
+const initialState = tree;
+
+const getAllDescendants = (tree, id) => {
+  const descendants = [id];
+  const run = (id) => {
```

Comment:
:hand: Hmm. Writing this as a recursive function _works_, but it's a bit unusual.  Is there another way to do this? Maybe writing it as a queue that we keep pushing into?

Reply:
You're right that it does look strange, but it's the fastest that I written so far https://jsbench.me/w0kgrpkfl0/1, it compares the previous one which was an array reduce, breadth first search (is that what you mean by writing as a queue).

Here's one where it pushes ids into a queue. From my online research, js implementations of BFS use array.shift, but it slows it down, so I used  a index to retrieve the value instead, so it's much faster but increases space complexity.
```js
const getAllDescendantIds = (tree, id) => {
  const queue = [id, ...tree[id].childIds]
  const descendants = []
  let indexPointer = 0

  while (indexPointer < queue.length) {
    const currentItem = queue[indexPointer]
    descendants.push(currentItem)
    queue.push(...tree[currentItem].childIds)
    indexPointer++
  }

  return descendants
}
```